doc/reference/index.rst

```diff
@@ -23,3 +23,214 @@ API Reference
+  thread to sleep
+* :ref:`api_term_no-wait` if a parameter to the function can prevent the
+  invoking thread from trying to sleep
+* :ref:`api_term_isr-ok` if the function can always be safely called from non-thread
+  (interrupt) context even if it may return an error in that case
+* :ref:`api_term_pre-kernel-ok` if the function can be safely called
+  before the kernel has been fully initialized, even if it may return an
+  error in that case
+* :ref:`api_term_async` if the function may return before the operation
+  it initializes is complete (i.e. function return and operation
+  completion are asynchronous)
+* :ref:`api_term_supervisor` if the calling thread must have supervisor
+  privileges to execute the function
+
+Details on the behavioral impact of each attribute are in the following
+sections.  Attributes for some common kernel API are below:
+
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|name                           | reschedule | sleep | no-wait | isr-ok | async | supervisor|
++===============================+============+=======+=========+========+=======+===========+
+|:c:func:`k_busy_wait`          | no         | no*   | no*     | yes*   | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_sem_give`           | yes        | no    | no*     | yes*   | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_sleep`              | yes*       | yes   | no      | no     | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_sem_take`           | yes*       | yes   | yes     | no     | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`k_thread_create`      | ?          | ?     | ?       | ?      | no    | yes       |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`spi_transceive`       | yes*       | yes   | no      | no     | no    | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+|:c:func:`spi_transceive_async` | yes*       | yes   | no      | no     | yes   | no        |
++-------------------------------+------------+-------+---------+--------+-------+-----------+
+
+``*`` indicates the attribute value is implicit from another attribute.
+For example a function that is not **reschedule** cannot be **sleep**;
+one that is **sleep** is necessarily **reschedule**; and one that cannot
+sleep has no use for no-wait support and is implicitly **isr-ok**.
+
+.. _api_term_reschedule:
+
+reschedule
+==========
+
+The reschedule attribute is used on a function that can reach a
+:ref:`reschedule point <scheduling_v2>` within its execution.
+
+.. note::
```

Comment:
I would drop the "note" and make the text continous, also in all following terms.

Reply:
I'm willing to do that to get this merged, but I'd like additional perspective before committing to it.  The material before the note is intended to be normative and technically precise.  The material in the note is intended to be commentary that is motivational and explanatory, saying why this is important, how to use it correctly, and how it affects use of the corresponding API.

If there's a different type of callout that should be used I'd be happy to switch to it.  @dbkinder ?

Reply:
Maybe use definition lists...

Reply:
Using a note directive causes the enclosed text to be particularly highlighted when rendered.  Given your comment that this "note" is commentary, it would render nicer if you just added a subheading, something like this:
```
reschedule
==========

The reschedule attribute is used on a function that can reach a
:ref:`reschedule point <scheduling_v2>` within its execution.

Explanation
-----------
The significance of this attribute is that when a rescheduling
function is invoked by a thread it is possible for that thread to be
suspended as a consequence of a higher-priority thread being made
...

```